# Part 3: Sequence Classification

__Before starting, we recommend you enable GPU acceleration if you're running on Colab.__

In [0]:
# Execute this code block to install dependencies when running on colab
try:
    import torch
except:
    from os.path import exists
    from wheel.pep425tags import get_abbr_impl, get_impl_ver, get_abi_tag
    platform = '{}{}-{}'.format(get_abbr_impl(), get_impl_ver(), get_abi_tag())
    cuda_output = !ldconfig -p|grep cudart.so|sed -e 's/.*\.\([0-9]*\)\.\([0-9]*\)$/cu\1\2/'
    accelerator = cuda_output[0] if exists('/dev/nvidia0') else 'cpu'

    !pip install -q http://download.pytorch.org/whl/{accelerator}/torch-1.0.0-{platform}-linux_x86_64.whl torchvision

try: 
    import torchbearer
except:
    !pip install torchbearer

try:
    import torchtext
except:
    !pip install torchtext
    
try:
    import spacy
except:
    !pip install spacy
    
try:
    spacy.load('en')
except:
    !python -m spacy download en

## Sequence Classification
The problem that we will use to demonstrate sequence classification in this lab is the IMDB movie review sentiment classification problem. Each movie review is a variable sequence of words and the sentiment of each movie review must be classified.

The Large Movie Review Dataset (often referred to as the IMDB dataset) contains 25,000 highly-polar movie reviews (good or bad) for training and the same amount again for testing. The problem is to determine whether a given movie review has a positive or negative sentiment. The data was collected by Stanford researchers and was used in a 2011 paper where a split of 50-50 of the data was used for training and test. An accuracy of 88.89% was achieved.

We'll be using a **recurrent neural network** (RNN) as they are commonly used in analysing sequences. An RNN takes in sequence of words, $X=\{x_1, ..., x_T\}$, one at a time, and produces a _hidden state_, $h$, for each word. We use the RNN _recurrently_ by feeding in the current word $x_t$ as well as the hidden state from the previous word, $h_{t-1}$, to produce the next hidden state, $h_t$. 

$$h_t = \text{RNN}(x_t, h_{t-1})$$

Once we have our final hidden state, $h_T$, (from feeding in the last word in the sequence, $x_T$) we feed it through a linear layer, $f$, (also known as a fully connected layer), to receive our predicted sentiment, $\hat{y} = f(h_T)$.

Below shows an example sentence, with the RNN predicting zero, which indicates a negative sentiment. The RNN is shown in orange and the linear layer shown in silver. Note that we use the same RNN for every word, i.e. it has the same parameters. The initial hidden state, $h_0$, is a tensor initialized to all zeros. 

![](http://comp6248.ecs.soton.ac.uk/labs/lab7/assets/sentiment1.png)

**Note:** some layers and steps have been omitted from the diagram, but these will be explained later.


The TorchText library provides easy access to the IMDB dataset. The `IMDB` class allows you to load the dataset in a format that is ready for use in neural network and deep learning models, and TorchText's utility methods allow us to easily create batches of data that are `padded` to the same length (we need to pad shorter sentences in the batch to the length of the longest sentence).

One of the main concepts of TorchText is the `Field`. These define how your data should be processed. In our sentiment classification task the data consists of both the raw string of the review and the sentiment, either "pos" or "neg".

The parameters of a `Field` specify how the data should be processed. 

We use the `TEXT` field to define how the review should be processed, and the `LABEL` field to process the sentiment. 

Our `TEXT` field has `tokenize='spacy'` as an argument. This defines that the "tokenization" (the act of splitting the string into discrete "tokens") should be done using the [spaCy](https://spacy.io) tokenizer. If no `tokenize` argument is passed, the default is simply splitting the string on spaces.

`LABEL` is defined by a `LabelField`, a special subset of the `Field` class specifically used for handling labels. We will explain the `dtype` argument later.

For more on `Fields`, go [here](https://github.com/pytorch/text/blob/master/torchtext/data/field.py).

In [0]:
import torch
from torchtext import data

TEXT = data.Field(tokenize='spacy', lower=True, include_lengths=True)
LABEL = data.LabelField(dtype=torch.float)

The following code automatically downloads the IMDb dataset and splits it into the canonical train/test splits as `torchtext.datasets` objects. It process the data using the `Fields` we have previously defined. Note that the following can take a couple of minutes to run due to the tokenisation:

In [0]:
from torchtext import datasets

train_data, test_data = datasets.IMDB.splits(TEXT, LABEL)

We can see how many examples are in each split by checking their length.

In [4]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 25000
Number of testing examples: 25000


We can also check an example.

In [5]:
print(vars(train_data.examples[0]))

{'text': ['with', 'all', 'that', 's', 'going', 'on', 'in', 'the', 'world', 'sometimes', 'we', 'need', 'an', 'escape', '.', 'curly', 'sue', 'is', 'just', 'that', '.', 'not', 'a', 'complicated', 'plot', 'or', 'deep', 'meaning', ';', 'however', 'it', 'is', 'not', 'devoid', 'of', 'substance', '.', 'there', 'is', 'more', 'than', 'furious', 'action', 'or', 'heart', 'pounding', 'dramas', '.', 'there', 'are', 'the', 'charming', 'little', 'shows', 'you', 'can', 'watch', 'with', 'your', 'kids', 'and', 'have', 'enough', 'substance', 'to', 'enjoy', 'with', 'your', 'date', '.', 'try', 'it', 'you', 'may', 'like', 'it', 'more', 'than', 'you', 'think', '.', 'the', 'little', 'girl', 'is', 'really', 'smart', 'and', 'cute', '.', 'the', '"', 'dad', '"', 'and', 'the', 'girl', 'go', 'thru', 'some', 'slapstick', 'routines', '.', 'when', 'a', 'jealous', 'boyfriend', 'steps', 'in', ',', 'trouble', 'brews', 'for', 'curly', 'and', 'the', 'life', 'she', 's', 'known', 'may', 'be', 'torn', 'asunder', '.', 'fred', '

The IMDb dataset only has train/test splits, so we need to create a validation set. We can do this with the `.split()` method. 

By default this splits 70/30, however by passing a `split_ratio` argument, we can change the ratio of the split, i.e. a `split_ratio` of 0.8 would mean 80% of the examples make up the training set and 20% make up the validation set. 

We also pass our random seed to the `random_state` argument, ensuring that we get the same train/validation split each time.

In [0]:
import random

train_data, valid_data = train_data.split()

Again, we'll view how many examples are in each split.

In [7]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of validation examples: {len(valid_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 17500
Number of validation examples: 7500
Number of testing examples: 25000


Next, we have to build a _vocabulary_. This is effectively a look up table where every unique word in your data set has a corresponding _index_ (an integer).

We do this as our machine learning model cannot operate on strings, only numbers. Each _index_ is used to construct a _one-hot_ vector for each word. A one-hot vector is a vector where all of the elements are 0, except one, which is 1, and dimensionality is the total number of unique words in your vocabulary, commonly denoted by $V$.

![](http://comp6248.ecs.soton.ac.uk/labs/lab7/assets/sentiment5.png)

The number of unique words in our training set is over 100,000, which means that our one-hot vectors will have over 100,000 dimensions! This will make training slow and possibly won't fit onto your GPU (if you're using one). 

There are two ways to effectively cut-down our vocabulary, we can either only take the top $n$ most common words or ignore words that appear less than $m$ times. We'll do the former, only keeping the top 25,000 words.

What do we do with words that appear in examples but we have cut from the vocabulary? We replace them with a special _unknown_ or `<unk>` token. For example, if the sentence was "This film is great and I love it" but the word "love" was not in the vocabulary, it would become "This film is great and I `<unk>` it".

The following builds the vocabulary, only keeping the most common `max_size` tokens.

In [0]:
"""
This makes torchtext go through all the elements in the training set, check the contents corresponding to the TEXT field, and register the words in its vocabulary. 
Torchtext has its own class called Vocab for handling the vocabulary. The Vocab class holds a mapping from word to id in its stoi attribute and a reverse mapping in its itos attribute. 
"""

TEXT.build_vocab(train_data, max_size=25000)
LABEL.build_vocab(train_data)

Why do we only build the vocabulary on the training set? When testing any machine learning system you do not want to look at the test set in any way. We do not include the validation set as we want it to reflect the test set as much as possible.

In [9]:
print(f"Unique tokens in TEXT vocabulary: {len(TEXT.vocab)}")
print(f"Unique tokens in LABEL vocabulary: {len(LABEL.vocab)}")

Unique tokens in TEXT vocabulary: 25002
Unique tokens in LABEL vocabulary: 2


Why is the vocab size 25002 and not 25000? One of the addition tokens is the `<unk>` token and the other is a `<pad>` token.

When we feed sentences into our model, we feed a _batch_ of them at a time, i.e. more than one at a time, and all sentences in the batch need to be the same size. Thus, to ensure each sentence in the batch is the same size, any sentences which are shorter than the longest within the batch are padded.

![](http://comp6248.ecs.soton.ac.uk/labs/lab7/assets/sentiment6.png)

We can also view the most common words in the vocabulary and their frequencies.

In [10]:
print(TEXT.vocab.freqs.most_common(20))

[('the', 230138), (',', 192761), ('.', 165326), ('and', 113897), ('a', 113489), ('of', 101928), ('to', 94751), ('is', 76915), ('it', 65420), ('in', 65066), ('i', 58285), ('this', 51249), ('that', 50894), ('"', 44049), ("'s", 43641), ('-', 37032), ('was', 35219), ('/><br', 35194), ('as', 32568), ('with', 30815)]


We can also see the vocabulary directly using either the `stoi` (**s**tring **to** **i**nt) or `itos` (**i**nt **to**  **s**tring) method.

In [11]:
print(TEXT.vocab.itos[:10])

['<unk>', '<pad>', 'the', ',', '.', 'and', 'a', 'of', 'to', 'is']


In [12]:
print(TEXT.vocab.stoi)

defaultdict(<function _default_unk_index at 0x7f14cd256158>, {'<unk>': 0, '<pad>': 1, 'the': 2, ',': 3, '.': 4, 'and': 5, 'a': 6, 'of': 7, 'to': 8, 'is': 9, 'it': 10, 'in': 11, 'i': 12, 'this': 13, 'that': 14, '"': 15, "'s": 16, '-': 17, 'was': 18, '/><br': 19, 'as': 20, 'with': 21, 'for': 22, 'movie': 23, 'but': 24, 'film': 25, 'on': 26, 'you': 27, '(': 28, ')': 29, "n't": 30, 'not': 31, 'are': 32, 'he': 33, 'his': 34, 'have': 35, 'be': 36, 'one': 37, 'all': 38, 'at': 39, 'they': 40, 'by': 41, '!': 42, 'an': 43, 'who': 44, 'from': 45, 'like': 46, 'so': 47, 'her': 48, 'just': 49, 'there': 50, 'or': 51, 'do': 52, "'": 53, 'about': 54, 'has': 55, 'out': 56, 'if': 57, 'what': 58, 'some': 59, 'good': 60, '?': 61, 'more': 62, 'very': 63, 'she': 64, 'when': 65, 'would': 66, 'up': 67, 'no': 68, 'even': 69, 'time': 70, 'can': 71, 'which': 72, 'my': 73, 'only': 74, 'story': 75, 'really': 76, 'had': 77, 'their': 78, 'see': 79, 'were': 80, 'we': 81, 'does': 82, 'did': 83, 'me': 84, 'well': 85, 't

We can also check the labels, ensuring 0 is for negative and 1 is for positive.

In [13]:
print(LABEL.vocab.stoi)

defaultdict(<function _default_unk_index at 0x7f14cd256158>, {'neg': 0, 'pos': 1})


The final step of preparing the data is creating the iterators. We iterate over these in the training/evaluation loop, and they return a batch of examples (indexed and converted into tensors) at each iteration.

We'll use a `BucketIterator` which is a special type of iterator that will return a batch of examples where each example is of a similar length, minimizing the amount of padding per example. Torchtext will pad for us automatically (handled by the `Field` object). We'll request the items within each batch produced by the `BucketIterator` are sorted by length.

We also want to place the tensors returned by the iterator on the GPU (if you're using one). PyTorch handles this using `torch.device`, we then pass this device to the iterator.

In [14]:
type(train_data)

torchtext.data.dataset.Dataset

In [0]:
BATCH_SIZE = 64

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

train_iterator, valid_iterator, test_iterator = data.BucketIterator.splits(
    (train_data, valid_data, test_data), 
    batch_size=BATCH_SIZE,
    device=device,
    sort_key=lambda x: len(x.text),
    sort_within_batch=True)

## Build the Model

The next stage is building the model that we'll eventually train and evaluate. 

There is a small amount of boilerplate code when creating models in PyTorch, note how our `RNN` class is a sub-class of `nn.Module` and the use of `super`.

Within the `__init__` we define the _layers_ of the module. Our three layers are an _embedding_ layer, our RNN, and a _linear_ layer. All layers have their parameters initialized to random values, unless explicitly specified.

The embedding layer is used to transform our sparse one-hot vector (sparse as most of the elements are 0) into a dense embedding vector (dense as the dimensionality is a lot smaller and all the elements are real numbers). This embedding layer is simply a single fully connected layer. As well as reducing the dimensionality of the input to the RNN, there is the theory that words which have similar impact on the sentiment of the review are mapped close together in this dense vector space. For more information about word embeddings, see [here](https://monkeylearn.com/blog/word-embeddings-transform-text-numbers/).

The RNN layer is our RNN which takes in our dense vector and the previous hidden state $h_{t-1}$, which it uses to calculate the next hidden state, $h_t$.

![](http://comp6248.ecs.soton.ac.uk/labs/lab7/assets/sentiment7.png)

Finally, the linear layer takes the final hidden state and feeds it through a fully connected layer, $f(h_T)$, transforming it to the correct output dimension.

The `forward` method is called when we feed examples into our model.

Each batch, `text_len`, is a tuple containing a tensor of size _**[max_sentence length, batch size]**_ and a tensor of **batch_size** containing the true lengths of each sentence (remember, they won't necessarily be the same; some reviews are much longer than others). 

The first tensor in the tuple contains the ordered word indexes for each review in the batch. The act of converting a list of tokens into a list of indexes is commonly called *numericalizing*.

The input batch is then passed through the embedding layer to get `embedded`, which gives us a dense vector representation of our sentences. `embedded` is a tensor of size _**[sentence length, batch size, embedding dim]**_. 

`embedded` is then fed into a function called `pack_padded_sequence` before being fed into the RNN. `pack_padded_sequence` is used to create a datastructure that allows the RNN to 'mask' off the padding during the BPTT process (we don't want to learn the padding, as this could drastically influence the results!). In some frameworks you must feed the initial hidden state, $h_0$, into the RNN, however in PyTorch, if no initial hidden state is passed as an argument it defaults to a tensor of all zeros.

The RNN returns 2 tensors, `output` of size _**[sentence length, batch size, hidden dim]**_ and `hidden` of size _**[1, batch size, hidden dim]**_. `output` is the concatenation of the hidden state from every time step, whereas `hidden` is simply the final hidden state. 

Finally, we feed the last hidden state, `hidden`, through the linear layer, `fc`, to produce a prediction. Note the `squeeze` method, which is used to remove a dimension of size 1. 

nn.Embedding holds a Tensor of dimension (vocab_size, vector_size), i.e. of the size of the vocabulary x the dimension of each vector embedding, and a method that does the lookup.

When you create an embedding layer, the Tensor is initialised randomly. It is only when you train it when this similarity between similar words should appear. Unless you have overwritten the values of the embedding with a previously trained model, like GloVe or Word2Vec, but that's another story.

So, once you have the embedding layer defined, and the vocabulary defined and encoded (i.e. assign a unique number to each word in the vocabulary) you can use the instance of the nn.Embedding class to get the corresponding embedding.

In [0]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):
        super().__init__()
        
        self.embedding = nn.Embedding(input_dim, embedding_dim)
        self.rnn = nn.RNN(embedding_dim, hidden_dim,batch_first=True)
        self.fc = nn.Linear(hidden_dim, output_dim)
        
    def forward(self, text_len):
        text, lengths = text_len
        
        embedded = self.embedding(text)
        embedded = nn.utils.rnn.pack_padded_sequence(embedded, lengths)
        packed_output, hidden = self.rnn(embedded)

        return self.fc(hidden.squeeze(0))

We now create an instance of our RNN class. 

The input dimension is the dimension of the one-hot vectors, which is equal to the vocabulary size. 

The embedding dimension is the size of the dense word vectors. This is usually around 50-250 dimensions, but depends on the size of the vocabulary.

The hidden dimension is the size of the hidden states. This is usually around 100-500 dimensions, but also depends on factors such as on the vocabulary size, the size of the dense vectors and the complexity of the task.

The output dimension is usually the number of classes, however in the case of only 2 classes the output value is between 0 and 1 and thus can be 1-dimensional, i.e. a single scalar real number.

In [17]:
TEXT.vocab

In [0]:
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 50
HIDDEN_DIM = 100
OUTPUT_DIM = 1

model = RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)

# Train the model

Now we'll set up the training and then train the model.

First, we'll create an optimizer. This is the algorithm we use to update the parameters of the module. Here, we'll use _stochastic gradient descent_ (SGD). The first argument is the parameters that will be updated by the optimizer, the second is the learning rate, i.e. how much we'll change the parameters by when we do a parameter update.

In [0]:
import torch.optim as optim

optimizer = optim.SGD(model.parameters(), lr=0.001)

Next, we'll define our loss function. In PyTorch this is commonly called a criterion. 

The loss function here is _binary cross entropy with logits_. 

Our model currently outputs an unbound real number. As our labels are either 0 or 1, we want to restrict the predictions to a number between 0 and 1. We do this using the _sigmoid_ function. 

We then use this this bound scalar to calculate the loss using binary cross entropy. 

The `BCEWithLogitsLoss` criterion carries out both the sigmoid and the binary cross entropy steps.

In [0]:
criterion = nn.BCEWithLogitsLoss()

Finally, before creating a Torchbearer trial and copying the model to the GPU (if we have one), we need to adapt the format of the batches returned by TorchText. TorchText's iterators return an object representing a batch of data from which you can access the fields used to store the underlying data (e.g. the labels and words in our case). Torchbearer on the other hand needs to know what the X's and y's are explicitly. 

There are a number of ways in which we can adapt the batch data format, but one of the easiest conceptually is to write a simple wrapper iterator that reads the next torchtext batch and returns a tuple of objects (the X and y). Note that because we specified that we want the lengths of the sentences earlier, the X's will be a tuple of the sentences and their lengths. If we had not requested the lengths, then the X's would just be a tensor encoding the padded sentences.

In [0]:
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
model = model.to(device)

In [0]:
from torchbearer import Trial
        
class MyIter:
    def __init__(self, it):
        self.it = it
    def __iter__(self):
        for batch in self.it:
            yield (batch.text, batch.label.unsqueeze(1))
    def __len__(self):
        return len(self.it)



In [0]:

import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

In [24]:
torchbearer_trial = Trial(model, optimizer, criterion, metrics=['acc', 'loss']).to(device)
torchbearer_trial.with_generators(train_generator=MyIter(train_iterator), val_generator=MyIter(valid_iterator), test_generator=MyIter(test_iterator))
torchbearer_trial.run(epochs=5)
torchbearer_trial.predict()

tensor([[-0.0055],
        [-0.2000],
        [ 0.1190],
        ...,
        [ 0.2109],
        [ 0.0772],
        [-0.2443]], device='cuda:0')

__Use the box below to comment on and give insight into the performance of the above model:__

The accuracy is as low as 50%, even on training for 10 or more epochs. Its just equivalent to random guessing of the sentiment of the review.

Now try and build a better model. Rather than using a plain RNN, we'll instead use a (single layer) LSTM, and we'll use Adam with an initial learning rate of 0.01 as the optimiser. __Complete the following code to implement the improved model, and then train it:__

In [0]:
class ImprovedRNN(nn.Module):
    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):
        super().__init__()
        
        self.embedding = nn.Embedding(input_dim, embedding_dim)
        # YOUR CODE HERE
        self.rnn = nn.LSTM(embedding_dim, hidden_dim)
        self.fc = nn.Linear(hidden_dim, output_dim)
        
    def forward(self, text_len):
        text, lengths = text_len
        embedded = self.embedding(text)
        embedded = nn.utils.rnn.pack_padded_sequence(embedded, lengths)
        packed_output, (hidden,cn) = self.rnn(embedded)
        return self.fc(hidden.squeeze(0))



In [26]:
imodel = ImprovedRNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)
newoptimizer = optim.Adam(imodel.parameters(), lr=0.01)
# TODO: Train and evaluate the model
# YOUR CODE HERE
torchbearer_trial = Trial(imodel, newoptimizer, criterion, metrics=['acc', 'loss']).to(device)
torchbearer_trial.with_generators(train_generator=MyIter(train_iterator), val_generator=MyIter(valid_iterator), test_generator=MyIter(test_iterator))
torchbearer_trial.run(epochs=10)
torchbearer_trial.predict()
#raise NotImplementedError()

tensor([[ -9.5500],
        [ -7.0419],
        [-12.4847],
        ...,
        [  7.3170],
        [ -1.5485],
        [ -0.5825]], device='cuda:0')

__What do you observe about the performance of this model? What would you do next if you wanted to improve it further? Write your answers in the box below:__

The performance of the improved model as well as any other model has almost doubled. Training accuracy reaches 99.6%, whereas the validation set and test set accuracy are 87% and 85% respectively.

But as step for improvement, we could 
1. add multiple layers, 
2. make them bidirectional
3. increase the size of the vocabulary and embedding.
4. add additional dense layers

Also, it is seen that while predicting sentiment with test reviews, 0.8454253673553467 is the score for 'Not a good movie', and 0.8632851243019104 is the score for 'good movie'. This tells us that the model still doesn't semantically understand the whole sentence. It does well for extreme reviews, like reviews with words like 'Great', 'horrible', and if the review has exclamation marks, are obvious to the model. But the neutral and subtle reviews are still a problem. 

In [27]:
|# YOUR CODE HERE
#raise NotImplementedError()
'''
class ImprovedRNN2(nn.Module):
    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):
        super().__init__()
        
        self.embedding = nn.Embedding(input_dim, embedding_dim*2)
        # YOUR CODE HERE
        self.rnn = nn.LSTM(embedding_dim*2, hidden_dim)
        self.fc1 = nn.Linear(hidden_dim, output_dim*2)
        self.fc2 = nn.Linear(output_dim*2,output_dim)
        
    def forward(self, text_len):
        text, lengths = text_len
        embedded = self.embedding(text)
        embedded = nn.utils.rnn.pack_padded_sequence(embedded, lengths)
        packed_output, (hidden,cn) = self.rnn(embedded)
        #hidden = torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim=1)
        out1 = self.fc1(hidden.squeeze(0))
        #out = self.fc2(out1)
        return self.fc2(out1.squeeze(0))
'''

'\nclass ImprovedRNN2(nn.Module):\n    def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):\n        super().__init__()\n        \n        self.embedding = nn.Embedding(input_dim, embedding_dim*2)\n        # YOUR CODE HERE\n        self.rnn = nn.LSTM(embedding_dim*2, hidden_dim)\n        self.fc1 = nn.Linear(hidden_dim, output_dim*2)\n        self.fc2 = nn.Linear(output_dim*2,output_dim)\n        \n    def forward(self, text_len):\n        text, lengths = text_len\n        embedded = self.embedding(text)\n        embedded = nn.utils.rnn.pack_padded_sequence(embedded, lengths)\n        packed_output, (hidden,cn) = self.rnn(embedded)\n        #hidden = torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim=1)\n        out1 = self.fc1(hidden.squeeze(0))\n        #out = self.fc2(out1)\n        return self.fc2(out1.squeeze(0))\n'

In [28]:
'''
model2 = ImprovedRNN2(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)
newoptimizer = optim.Adam(model2.parameters(), lr=0.01)
# TODO: Train and evaluate the model
# YOUR CODE HERE
torchbearer_trial = Trial(model2, newoptimizer, criterion, metrics=['acc', 'loss']).to(device)
torchbearer_trial.with_generators(train_generator=MyIter(train_iterator), val_generator=MyIter(valid_iterator), test_generator=MyIter(test_iterator))
torchbearer_trial.run(epochs=10)
torchbearer_trial.predict()
#raise NotImplementedError()

'''

"\nmodel2 = ImprovedRNN2(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)\nnewoptimizer = optim.Adam(model2.parameters(), lr=0.01)\n# TODO: Train and evaluate the model\n# YOUR CODE HERE\ntorchbearer_trial = Trial(model2, newoptimizer, criterion, metrics=['acc', 'loss']).to(device)\ntorchbearer_trial.with_generators(train_generator=MyIter(train_iterator), val_generator=MyIter(valid_iterator), test_generator=MyIter(test_iterator))\ntorchbearer_trial.run(epochs=10)\ntorchbearer_trial.predict()\n#raise NotImplementedError()\n\n"

## User Input

We can now use our models to predict the sentiment of any sentence we give it. As it has been trained on movie reviews, the sentences provided should also be movie reviews.

Our `predict_sentiment` function does a few things:
- tokenizes the sentence, i.e. splits it from a raw string into a list of tokens
- indexes the tokens by converting them into their integer representation from our vocabulary
- converts the indexes, which are a Python list into a PyTorch tensor
- add a batch dimension by `unsqueeze`ing 
- squashes the output prediction from a real number between 0 and 1 with the `sigmoid` function
- converts the tensor holding a single value into an integer with the `item()` method

We are expecting reviews with a negative sentiment to return a value close to 0 and positive reviews to return a value close to 1.

In [0]:
import spacy
nlp = spacy.load('en')

def predict_sentiment(imodel, sentence):
    tokenized = [tok.text for tok in nlp.tokenizer(sentence)]
    indexed = [TEXT.vocab.stoi[t] for t in tokenized]
    tensor = torch.LongTensor(indexed).to(device)
    tensor = tensor.unsqueeze(1)
    prediction = torch.sigmoid(imodel((tensor, torch.tensor([tensor.shape[0]]))))
    return prediction.item()

An example negative review...

In [39]:
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
imodel = imodel.to(device)
predict_sentiment(imodel, "This film is terrible")

9.925320046022534e-05

and an example positive review...

In [46]:
predict_sentiment(imodel, "This film is good")

0.9781243801116943

__Use the box below to try classifying some of your own 'movie reviews':__

In [41]:
predict_sentiment(imodel, "This movie made me cry")

0.945192277431488

In [42]:
predict_sentiment(imodel, "worst ever")

0.0013021227205172181

In [43]:
predict_sentiment(imodel, "bad bad bad")

0.010986844077706337

In [44]:
predict_sentiment(imodel, "I'm not sure how this movie is")

0.00023457311908714473

In [55]:
predict_sentiment(imodel, "Not a good movie")

0.8454253673553467

In [57]:
predict_sentiment(imodel, "good movie!")

0.9038721919059753